lib/galaxy/webapps/galaxy/controllers/history.py

```diff
@@ -1237,59 +1240,49 @@ def set_accessible_async(self, trans, id=None, accessible=False):
         return
         # TODO: used in page/editor.mako
 
-    @web.expose
+    @web.expose_api
     @web.require_login("rename histories")
-    def rename(self, trans, id=None, name=None, **kwd):
-        user = trans.get_user()
+    def rename(self, trans, payload=None, **kwd):
+        id = kwd.get('id')
         if not id:
-            # Default to the current history
-            history = trans.get_history()
-            if not history.user:
-                return trans.show_error_message("You must save your history before renaming it.")
-            id = trans.security.encode_id(history.id)
-        id = galaxy.util.listify(id)
-        name = galaxy.util.listify(name)
+            return message_exception(trans, 'No history id received for renaming.')
+        user = trans.get_user()
+        id = listify(id)
         histories = []
-
         for history_id in id:
             history = self.history_manager.get_owned(self.decode_id(history_id), trans.user, current_history=trans.history)
             if history and history.user_id == user.id:
                 histories.append(history)
-        if not name or len(histories) != len(name):
-            return trans.fill_template("/history/rename.mako", histories=histories)
-
-        change_msgs = []
-        for i in range(len(histories)):
-            cur_name = histories[i].get_display_name()
-            new_name = name[i]
-
-            # skip if name is empty
-            if not isinstance(new_name, string_types) or not new_name.strip():
-                change_msgs.append("You must specify a valid name for History: " + cur_name)
-                continue
-
-            # skip if not the owner
-            # ??: isn't this already handled in get_history/if statement above?
-            if histories[i].user_id != user.id:
-                change_msgs.append("History: " + cur_name + " does not appear to belong to you.")
-                continue
-
-            # skip if it wouldn't be a change
-            if new_name == cur_name:
-                change_msgs.append("History: " + cur_name + " is already named: " + new_name)
-                continue
-
-            # escape, sanitize, set, and log the change
-            new_name = escape(new_name)
-            histories[i].name = sanitize_html(new_name)
-            trans.sa_session.add(histories[i])
-            trans.sa_session.flush()
-
-            trans.log_event("History renamed: id: %s, renamed to: '%s'" % (str(histories[i].id), new_name))
-            change_msgs.append("History: " + cur_name + " renamed to: " + new_name)
-
-        change_msg = '<br />'.join(change_msgs)
-        return trans.show_message(change_msg, refresh_frames=['history'])
+        if trans.request.method == 'GET':
+            return {
+                'title'  : 'Change history name(s)',
+                'inputs' : [{
+                    'name'  : trans.security.encode_id(h.id),
```

Comment:
This seems like a significant regression in terms of... I don't know what the word is... I guess navigability of the DOM. The DOM is not as semantically useful as it was previously, I'll merge this but I really hope you consider these issues going forward. It sets a "tour_id" (an unfortunate name given that it has many applications besides tours including tests and other automated driving of the UI) but the ID isn't something that can be predicted for tours and it isn't something intuitive or stable across histories. There is no ``form`` element for the ``form`` (there should probably be one IMO - [xref](https://stackoverflow.com/a/6309374)). Hitting enter no longer submits the form like it did before. To find the input itself I had to use ``'.ui-form-element input.ui-input'`` instead of the more intuitive ``input[name="name"]`` where it is clear what the input is for (the name).

<img width="557" alt="screen shot 2017-09-26 at 9 09 28 am" src="https://user-images.githubusercontent.com/216771/30863106-297a1e64-a29e-11e7-98fa-a5203594685c.png">

I don't mind all the extra divs and such - as long as semantically intuitive selectors can be constructed.

Reply:
This is something we've argued about across several issues; I'll second (again) the concern here.  "tour_id" should not be a necessary construct, when we already *have* a language for describing identifiers on the page.  We just need to use it better.

Reply:
Thx for the review. I modified the input element names in #4700 by relying on enumeration. The resulting 'tour_id's are 'name_INDEX' instead of the encoded history id.